tests/cli/shm_ui.c

```diff
@@ -0,0 +1,471 @@
+        // Get shm subscriptions, command values, and data values
+        get_sub_requests(sub_map_net_handler_all_devs, NET_HANDLER);    // Get subscriptions made by net handler
+        get_sub_requests(sub_map_executor_all_devs, EXECUTOR);          // Get subscriptions made by executor
+        get_cmd_map(cmd_map_all_devs);
+        device_read(shm_idx, SHM, COMMAND, ~0, command_vals);
+        device_read(shm_idx, SHM, DATA, ~0, data_vals);
+    }
+    
+    // We care about only the specified device (this is just for the sake of brevity)
+    uint32_t sub_map_net_handler = sub_map_net_handler_all_devs[shm_idx + 1];   
+    uint32_t sub_map_executor = sub_map_executor_all_devs[shm_idx + 1];
+    uint32_t cmd_map = cmd_map_all_devs[shm_idx + 1];
+    
+    // Each iteration prints out a row for each parameter
+    int display_cmd_val; // Flag for whether we should display the command value for a parameter
+    for (int i = 0; i < num_params; i++) {
+        wclrtoeol(DEVICE_WIN); // Clear previous parameter entry
+
+        // Print the index and its name
+        mvwprintw(DEVICE_WIN, line, PARAM_IDX_COL, "%d", i);
+        mvwprintw(DEVICE_WIN, line, PARAM_NAME_COL, "%s", show_custom_data ? custom_param_names[i] : device->params[i].name);
+
+        // Print out subscriptions ("X" iff subbed)
+        // Note that subscriptions are irrelevant for custom data
+        //   Runtime will *always* send custom data to net handler
+        //   Executor has nothing to do with the custom data
+        if (show_custom_data || sub_map_net_handler & (1 << i)) {
+            mvwprintw(DEVICE_WIN, line, NET_SUB_COL + 1, "X");
+        }
+        if (!show_custom_data && sub_map_executor & (1 << i)) {
+            mvwprintw(DEVICE_WIN, line, EXE_SUB_COL + 1, "X");
+        }
+
+        // Determine whether we need to display the command value
+        if (show_custom_data) {
+            mvwprintw(DEVICE_WIN, line, COMMAND_VAL_COL, "N/A");
+            display_cmd_val = 0;
+        } else if (device->params[i].write == 0) { // Read-only parameter
+            mvwprintw(DEVICE_WIN, line, COMMAND_VAL_COL, "RD_ONLY");
+            display_cmd_val = 0;
+        } else if ((cmd_map & (1 << i)) == 0) { // No command to this parameter
+            mvwprintw(DEVICE_WIN, line, COMMAND_VAL_COL, "NONE");
+            display_cmd_val = 0;
+        } else { // There is a command for the write-able parameter
+            display_cmd_val = 1;
+        }
+
+        // Display values according to the parameter's type
+        param_type_t param_type = (show_custom_data) ? custom_param_types[i] : device->params[i].type;
+        switch (param_type) {
+            case INT:
+                if (display_cmd_val) {
+                    mvwprintw(DEVICE_WIN, line, COMMAND_VAL_COL, "%d", command_vals[i].p_i);
+                }
+                mvwprintw(DEVICE_WIN, line, DATA_VAL_COL, "%d", (show_custom_data) ? custom_param_values[i].p_i : data_vals[i].p_i);
```

Comment:
One way to reduce code is to move the ternary condition before the switch to get the `param_val_t`, similarly to what you did for the param type. Then you can just use that one struct in each of the 3 cases